ParameterVector support for pulse parameter assignment

arthurostrauss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
This PR proposes an extended use of the current method `assign_parameters` in the `ParameterManager`to enable the assignment of a list of values through `ParameterVector`class. This enables an extended usage for the same methods in the `Schedule`and `ScheduleBlock`objects.


### Details and comments

The PR simply takes the already existing unrolling method from `QuantumCircuit`and applies it for the pulse level case, to enable a simple extension of the existing methods for parameter assignment.
This can be practical when mixing pulse level calibrations with many parameters.
This PR closes #5233

arthurostrauss committed:
Added test and release notes update

arthurostrauss committed:
Added test and release notes update

arthurostrauss committed:
Added test and release notes update

arthurostrauss committed:
Added test and release notes update

arthurostrauss committed:
Reformatting

arthurostrauss committed:
Reformatting

arthurostrauss committed:
Reformatting

arthurostrauss committed:
Reformatting

arthurostrauss committed:
Merge branch 'main' of https://github.com/arthurostrauss/qiskit

arthurostrauss committed:
Added compatibility of pulse ParameterManager with ParameterVector

It is now possible to assign to a pulse schedule parameters in the form of a list of values that can be directly binded to ParameterVector. This PR is based on the current functioning of the analogous method for the QuantumCircuit class.

arthurostrauss committed:
Added compatibility of pulse ParameterManager with ParameterVector

It is now possible to assign to a pulse schedule parameters in the form of a list of values that can be directly binded to ParameterVector. This PR is based on the current functioning of the analogous method for the QuantumCircuit class.

arthurostrauss commented:
This PR would solve issue #5233

wshanks commented on qiskit/pulse/parameter_manager.py:
```diff
@@ -387,3 +393,38 @@ def update_parameter_table(self, new_node: Any):
+        out = {}
+        for parameter, value in parameter_binds.items():
+            if isinstance(parameter, ParameterVector):
+                if not isinstance(value, (list, tuple)):
```
Why does it check for `tuple` if the type hint only lists `list`? Perhaps everything should just be `collections.abc.Sequence`?

arthurostrauss replied:
The type is now typing.Sequence.

wshanks commented on releasenotes/notes/pulse_parameter_manager_compat_with_ParameterVector-7d31395fd4019827.yaml:
```diff
@@ -0,0 +1,14 @@
+---
+prelude: >
+    Compatibility of pulse parameter assignment with `ParameterVector` class.
```
I usually leave it to the release manager to decide what makes it into the prelude.

```suggestion
```

wshanks commented on qiskit/pulse/parameter_manager.py:
```diff
@@ -360,7 +361,9 @@ def get_parameters(self, parameter_name: str) -> list[Parameter]:
         pulse_program: Any,
-        value_dict: dict[ParameterExpression, ParameterValueType],
+        value_dict: dict[
+            ParameterExpression | ParameterVector, ParameterValueType | list[ParameterValueType]
```
Too bad there is no clean way to express paired key-value types on dicts.

wshanks commented on releasenotes/notes/pulse_parameter_manager_compat_with_ParameterVector-7d31395fd4019827.yaml:
```diff
@@ -0,0 +1,14 @@
+---
+prelude: >
+    Compatibility of pulse parameter assignment with `ParameterVector` class.
+features:
```
```suggestion
features_pulse:
```

wshanks commented on releasenotes/notes/pulse_parameter_manager_compat_with_ParameterVector-7d31395fd4019827.yaml:
```diff
@@ -0,0 +1,14 @@
+    An additional method `_unroll_param_dict`has been added to the `ParameterManager` class to unroll the
+    parameter dictionary into a list of parameter values. This method is used to convert the parameter dictionary
+    into a list of parameter values when using the `ParameterVector` class. This method is a replication of the
+    analogous method `_unroll_param_dict` in `QuantumCircuit` class.
```
This is good to note in the PR but doesn't need to be in the release notes since it is an internal helper function and the release notes are user facing.

```suggestion
```

wshanks commented on qiskit/pulse/schedule.py:
```diff
@@ -711,13 +712,18 @@ def is_parameterized(self) -> bool:
         Args:
-            value_dict: A mapping from Parameters to either numeric values or another
-                Parameter expression.
+            value_dict: A mapping from Parameters (ParameterVectors) to either
```
It was already using "Parameters" but I think here I would use "parameters" and "parameter vectors" unless you want to use Sphinx references to make it like `` :class:`.Parameter` ``s

wshanks reviewed (commented):
This looks great! I had some feedback on the documentation, but the code looks fine.

arthurostrauss committed:
Commit requested changes

Type for submitting a list of parameters has been set to Sequence for the case of ParameterVector. This enables the user to also pass a tuple of values/ParameterExpressions

arthurostrauss committed:
Commit requested changes

Type for submitting a list of parameters has been set to Sequence for the case of ParameterVector. This enables the user to also pass a tuple of values/ParameterExpressions

TsafrirA commented on releasenotes/notes/pulse_parameter_manager_compat_with_ParameterVector-7d31395fd4019827.yaml:
```diff
@@ -0,0 +1,10 @@
+features_pulse:
+  - |
+    The `ParameterManager` now supports the `ParameterVector` class. This enables
+    the use of `ParameterVector` in pulse `Schedule`and `ScheduleBlock` objects when using their method
```
```suggestion
    the use of `ParameterVector` in pulse `Schedule` and `ScheduleBlock` objects when using their method
```

TsafrirA commented on test/python/pulse/test_parameter_manager.py:
```diff
@@ -483,6 +483,22 @@ def test_parametric_pulses(self):
+        self.assertEqual(block.blocks[0].pulse.amp, 0.2)
+        self.assertEqual(block.blocks[0].pulse.sigma, 4.0)
+        self.assertEqual(block.blocks[1].phase, 0.1)
+
```
For completeness, I think there should be a similar test with `Schedule`.

TsafrirA reviewed (changes requested):
Looks good to me. Just a minor typo and perhaps an extra test.

arthurostrauss committed:
Update pulse_parameter_manager_compat_with_ParameterVector-7d31395fd4019827.yaml

Co-authored-by: TsafrirA <113579969+TsafrirA@users.noreply.github.com>

arthurostrauss committed:
Update pulse_parameter_manager_compat_with_ParameterVector-7d31395fd4019827.yaml

Co-authored-by: TsafrirA <113579969+TsafrirA@users.noreply.github.com>

arthurostrauss committed:
Added tests for Schedule and case of mix Parameter+numeric values

Complementary tests have been added for checking the pulse.Schedule.assign_parameters() method, as well as the functioning of binding to a ParameterVector a collection of numeric values and new ParameterExpression (through Parameter).

arthurostrauss committed:
Added tests for Schedule and case of mix Parameter+numeric values

Complementary tests have been added for checking the pulse.Schedule.assign_parameters() method, as well as the functioning of binding to a ParameterVector a collection of numeric values and new ParameterExpression (through Parameter).

arthurostrauss committed:
Corrected error for schedule test

An error was occurring when trying to get access to an instruction parameter within schedule.instructions

arthurostrauss committed:
Corrected error for schedule test

An error was occurring when trying to get access to an instruction parameter within schedule.instructions

arthurostrauss committed:
Correction to test schedule

arthurostrauss committed:
Correction to test schedule

arthurostrauss committed:
Merge branch 'ParameterVector_pulse' of https://github.com/arthurostrauss/qiskit into ParameterVector_pulse

TsafrirA commented:
Thanks @arthurostrauss for the final fixes! This looks great.
@nkanazawa1989 - Why commits from other PRs are marked as differences here?

wshanks commented on releasenotes/notes/pulse_parameter_manager_compat_with_ParameterVector-7d31395fd4019827.yaml:
```diff
@@ -0,0 +1,10 @@
+    the use of `ParameterVector` in pulse `Schedule` and `ScheduleBlock` objects when using their method
+    `assign_parameters` to assign simultaneously a list of parameter values.
+
+
```
This needs to be formatted as ReStructuredText rather than markdown. Here is I would phrase since it is a user-facing note and users do not generally work with `ParameterManager` directly.

```suggestion
features_pulse:
  - |
    The ``assign_parameters`` methods of :class:`.Schedule` and :class:`.ScheduleBlock`
    now support assigning a :class:`.ParameterVector` to a list of parameter values
    simultaneously in addition to assigning individual :class:`.Parameter` instances to
    individual values.
```

wshanks commented:
I think GitHub is confused about the multiple branches/merges. When I look at the diff locally (like `git checkout -b tmp Qiskit/main` and `git merge --squash arthurostrauss/ParameterVector_pulse`), it looks right.

arthurostrauss committed:
Update releasenotes/notes/pulse_parameter_manager_compat_with_ParameterVector-7d31395fd4019827.yaml

Co-authored-by: Will Shanks <wshaos@posteo.net>

wshanks commented:
@arthurostrauss This looks good to me now! Do you want to try to make it look clean in GitHub, so we know it will merge the way we want? I can squash things down and force push to your branch if you want.

arthurostrauss committed:
Merge branch 'main' into ParameterVector_pulse

arthurostrauss commented:
> @arthurostrauss This looks good to me now! Do you want to try to make it look clean in GitHub, so we know it will merge the way we want? I can squash things down and force push to your branch if you want.

@wshanks Yes if you can do such force push (if necessary) I'd be happy to let you do it, as I'm not sure how to proceed on my side for this.

wshanks commented:
@arthurostrauss Actually, the diff looks good in GitHub now after the last merge, so I think it is fine to merge as is. I am not sure what messed up the diff view before.
